Start iscsiuio only in presence of the relevant hardware (bsc#1194432)

ancorgs opened the pull request:
## Problem

According to [bsc#1194432](https://bugzilla.suse.com/show_bug.cgi?id=1194432) YaST enables the `iscsiuio` daemon always when some iSCSI devices are configured. But `iscsiuio` is the userspace offloading daemon for QLogic (formerly Broadcom) devices, so it's only needed when hardware offloading is configured for a device using the `bnx2i` or `qedi` drivers. In any other case, enabling `iscsiuio` only makes the boot process slower with no gain.

On the other hand, the convoluted code of yast-iscsi-client makes very hard to understand and modify the management of services.

## Solution

This pull request includes a couple of preliminary cleanup commits that should not change how the module behave. Instead, those commits provide small code cleanups and a lot of inline documentation to help to understand how services are handled in different parts of the module.

This pull request also includes a subsequent commit ensuring the `iscsiuio` service and socket are only enabled/disabled if there is any card in the system that should be handled with the `bnx2i` or `qedi` kernel modules. The implemented logic does not take into account whether hardware offloading is configured for such devices or not, so it still could enable the service in situations in which is not strictly needed. Nevertheless, the change should help in many scenarios and it should imply a low risk to introduce regressions.

The way services are started/stopped is not modified, only how they are enabled/disabled.

## Testing

Existing unit tests should be adapted and new unit tests should be created.

So far, that's left as an exercise to whoever finishes this pull request.

ancorgs commented on src/clients/iscsi-client_finish.rb:
```diff
@@ -101,6 +98,21 @@ def main
+      Builtins.y2milestone("enabling iscsiuio socket and service")
+      socket = Yast2::Systemd::Socket.find("iscsiuio")
+      socket.enable if socket
+      Service.Enable("iscsiuio")
```
@gonzoleeman You have permissions to add commits to this pull request. Feel free to add a commit to remove that line.

And also commits to adapt and extend the RSpec testsuite. Or do you want me to do so? I see it as a nice learning opportunity but we can also do it in a pair-programming session if you feel that would be better.

gonzoleeman replied:
> @gonzoleeman You have permissions to add commits to this pull request. Feel free to add a commit to remove that line.

Will do.

> 
> And also commits to adapt and extend the RSpec testsuite. Or do you want me to do so? I see it as a nice learning opportunity but we can also do it in a pair-programming session if you feel that would be better.

I would like to do that, but I don't know anything about RSpec, though I found this: [Relishapp.com link](https://relishapp.com/rspec) Do I download this package to get it: `ruby2.5-rubygem-rspec`

Do you have any documentation or advice on how to run it? I can hardly add new tests if I can't run them. :)

gonzoleeman commented:
I removed the service-enable for iscsiuio, and pushed it to this "iscsiuio" branch. I'm nervous because I can't test this myself, and I don't yet know how to run the self tests. But that's a separate issue.

ancorgs commented:
I will install a vanilla openSUSE Leap 15.3 from scratch and will configure it to run the RSpec tests, following (or improving) the currently available documentation and making a summary for you.

ancorgs commented:
I just checked and I can confirm that running the unit tests in a fresh-installed Leap 15.3 is as easy as:

```
sudo zypper in yast2-devtools ruby2.5-rubygem-yast-rake ruby2.5-rubygem-rspec
cd yast-iscsi-client
rake test:unit
```

The output clearly points which tests are failing. We need to fix them before this pull request can be considered acceptable.

ancorgs commented:
See? The unit tests were revealing a real bug introduced (by me) in this pull request. Fixed by [db061a](https://github.com/yast/yast-iscsi-client/pull/116/commits/db061ab2787add1e4a845f8986871b5acd6cab01).

So the tests are now adapted to one of the changes and they pass. But I would still consider this as an opportunity to extend the current testsuite to also cover the other parts we changed (like the Write part of `iscsi-client-finish` or the logic on `iscsiuio_relevant?`

@gonzoleeman Now that you have the mini-tutorial to setup RSpec and some commits as examples. Do you dare to try extending the testsuite? Maybe with some pair-programming session?

gonzoleeman commented:
> See? The unit tests were revealing a real bug introduced (by me) in this pull request. Fixed by [ffedaa7](https://github.com/yast/yast-iscsi-client/pull/116/commits/ffedaa78f3241b5655c9536f20a5b3221b4b13ae).
> 
> So the tests are now adapted to one of the changes and they pass. But I would still consider this as an opportunity to extend the current testsuite to also cover the other parts we changed (like the Write part of `iscsi-client-finish` or the logic on `iscsiuio_relevant?`).
> 
> @gonzoleeman Now that you have the mini-tutorial to setup RSpec and some commits as examples. Do you dare to try extending the testsuite? Maybe with some pair-programming session?

Got it. I will work on that asap. Apologies for not replying sooner, but I was working on a P1 that sucked all my time.

ancorgs commented:
> Got it. I will work on that asap. Apologies for not replying sooner, but I was working on a P1 that sucked all my time.

The good thing is that we are not in a hurry. Take your time and ask for help as often as needed.

gonzoleeman commented:
> > Got it. I will work on that asap. Apologies for not replying sooner, but I was working on a P1 that sucked all my time.
> 
> The good thing is that we are not in a hurry. Take your time and ask for help as often as needed.

I think this will be a great idea. Can we try to plan something for later next week? I'm attending LSF/MM next Mon/Tue/Wed, and I'm busy now madly preparing for that.

ancorgs commented:
> I think this will be a great idea. Can we try to plan something for later next week? I'm attending LSF/MM next Mon/Tue/Wed, and I'm busy now madly preparing for that.

Ok. I sent you an invitation for next Thursday.

dgdavid reviewed (approved):
@ancorgs, @gonzoleeman 


Looks good, thank you!

gonzoleeman commented:
I have pushed a test-banch to address this issue ... hoping my attempt is at least close to correct. Please review branch iscsiuio_fixes_maybe @ancorgs @dgdavid
